e2etest/zt_newe2e_oauth_device_test.go

```diff
@@ -0,0 +1,118 @@
+package e2etest
+
+import (
+	"flag"
+	"github.com/Azure/azure-storage-azcopy/v10/common"
+	"github.com/Azure/azure-storage-azcopy/v10/testSuite/cmd"
+	"path"
+	"strings"
+	"time"
+)
+
+var runDeviceCodeTest = flag.Bool("device-code", false, "Whether or not to run device code tests. These must be run manually due to interactive nature.")
+
+var manualEnv AzCopyEnvironment
+
+func init() {
+	if runDeviceCodeTest != nil && *runDeviceCodeTest {
+		suiteManager.RegisterSuite(&DeviceLoginManualSuite{})
+	}
+}
+
+func Scenario_CopySync(svm *ScenarioVariationManager, env *AzCopyEnvironment) {
+	azCopyVerb := ResolveVariation(svm, []AzCopyVerb{AzCopyVerbCopy, AzCopyVerbSync}) // Calculate verb early to create the destination object early
+	// Scale up from service to object
+	dstObj := CreateResource[ContainerResourceManager](svm, GetRootResource(svm, ResolveVariation(svm, []common.Location{common.ELocation.Local(), common.ELocation.Blob()})), ResourceDefinitionContainer{}).GetObject(svm, "test", common.EEntityType.File())
+	// The object must exist already if we're syncing.
+	if azCopyVerb == AzCopyVerbSync {
+		dstObj.Create(svm, NewZeroObjectContentContainer(0), ObjectProperties{})
+
+		if !svm.Dryrun() {
+			// Make sure the LMT is in the past
+			time.Sleep(time.Second * 10)
+		}
+	}
+
+	body := NewRandomObjectContentContainer(svm, SizeFromString("10K"))
+	// Scale up from service to object
+	srcObj := CreateResource[ObjectResourceManager](svm, GetRootResource(svm, ResolveVariation(svm, []common.Location{common.ELocation.Local(), common.ELocation.Blob()})), ResourceDefinitionObject{
+		ObjectName: pointerTo("test"),
+		Body:       body,
+	})
+
+	// no local->local
+	if srcObj.Location().IsLocal() && dstObj.Location().IsLocal() {
+		svm.InvalidateScenario()
+		return
+	}
+	azcopyLogPathFolder := cmd.GetAzCopyAppPath()
+	azcopyJobPlanFolder := path.Join(azcopyLogPathFolder, "plans")
+
+	stdout, _ := RunAzCopy(
+		svm,
+		AzCopyCommand{
+			Verb: azCopyVerb,
+			Environment: &AzCopyEnvironment{
+				ManualLogin:     true,
+				LogLocation:     &azcopyLogPathFolder,
+				JobPlanLocation: &azcopyJobPlanFolder,
+			},
+			Targets: []ResourceManager{
+				TryApplySpecificAuthType(srcObj, EExplicitCredentialType.OAuth(), svm, CreateAzCopyTargetOptions{}),
+				TryApplySpecificAuthType(dstObj, EExplicitCredentialType.OAuth(), svm, CreateAzCopyTargetOptions{})},
+			Flags: CopyFlags{
+				CopySyncCommonFlags: CopySyncCommonFlags{
+					Recursive: pointerTo(true),
+				},
+			},
+		})
+
+	ValidateResource[ObjectResourceManager](svm, dstObj, ResourceDefinitionObject{
+		Body: body,
+	}, true)
+
+	ValidateAADAuth(svm, stdout, srcObj.Location(), dstObj.Location())
+}
+
+func ValidateAADAuth(a Asserter, stdout AzCopyStdout, src, dst common.Location) {
```

Comment:
These also have a code smell about them-- I wanted to say we have a way to validate that stdout contains certain lines already, but, this is probably fine.

Reply:
Forgot about that, will fix to use that instead.